### PR TITLE
CI: Use conda compiler in `environment.yml`

### PR DESCRIPTION
**Pull request by marcelotrevisani**
Improvement to the conda environment to use the compiler from the virtual env instead of the system

Using the ``cxx-compile`` would be better because it will use the compiler from the virtual environment and it will not rely on the system anymore. That would be great to also avoid any kind of abi problems if someone uses some old compiler

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -10,7 +10,7 @@ clean_pyc:
 	-find . -name '*.py[co]' -exec rm {} \;
 
 build: clean_pyc
-	python setup.py build_ext --inplace
+	python setup.py build_ext --inplace --compiler=$CC
```
Comment (jreback) [inline Makefile]: this is not used in our CI at all, though I do use it locally, so nice to improve this generally.

We should add some documentation (in contributing.rst) that things exist & work (not 100% sure this actually works on windows; the Makefile)

Comment (marcelotrevisani) [reply Makefile]: That is true, unless ``pandas`` is using ``mingw`` it will not going to use it, indeed
but this part I am just enforcing it, because ``cython`` usually already uses the ``$CC``

**Comment by marcelotrevisani** [comment]
https://github.com/pandas-dev/pandas/pull/27061 should be merged first, after that I can rebase from ``main``

**Comment by marcelotrevisani** [comment]
For example, it will use the compiler from the conda virtual environment
```
/home/trevisani/miniconda3/envs/pandas-dev/bin/x86_64-conda_cos6-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -Wstrict-prototypes -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -fPIC -DNPY_NO_DEPRECATED_API=0 -I/home/trevisani/miniconda3/envs/pandas-dev/lib/python3.7/site-packages/numpy/core/include -I/home/trevisani/miniconda3/envs/pandas-dev/include/python3.7m -c pandas/util/move.c -o build/temp.linux-x86_64-3.7/pandas/util/move.o -Wno-unused-function
x86_64-conda_cos6-linux-gnu-gcc -pthread -shared -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,-rpath,/home/trevisani/miniconda3/envs/pandas-dev/lib -L/home/trevisani/miniconda3/envs/pandas-dev/lib -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,-rpath,/home/trevisani/miniconda3/envs/pandas-dev/lib -L/home/trevisani/miniconda3/envs/pandas-dev/lib -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 build/temp.linux-x86_64-3.7/pandas/util/move.o -o /home/trevisani/pandas/pandas/pandas/util/_move.cpython-37m-x86_64-linux-gnu.so

```

**Comment by marcelotrevisani** [comment]
Still need some work, the CI is getting the compiler from the system, at least for ``osx``

File: environment.yml [hunk 2]
```diff
@@ -79,5 +81,4 @@ dependencies:
   - xlrd  # pandas.read_excel, DataFrame.to_excel, pandas.ExcelWriter, pandas.ExcelFile
   - xlsxwriter  # pandas.read_excel, DataFrame.to_excel, pandas.ExcelWriter, pandas.ExcelFile
   - xlwt  # pandas.read_excel, DataFrame.to_excel, pandas.ExcelWriter, pandas.ExcelFile
-  - pip:
-    - pyreadstat  # pandas.read_spss
+  - pyreadstat  # pandas.read_spss
```
Comment (datapythonista) [inline environment.yml]: Can you revert this change here please. After that, you should run `./scripts/generate_pip_deps_from_conda.py` so `requirements.txt` is synchronized with the changes here.

File: environment.yml [hunk 3]
```diff
@@ -14,6 +14,8 @@ dependencies:
 
   # building
   - cython>=0.28.2
+  - cxx-compiler
+  - zlib
```
Comment (datapythonista) [inline environment.yml]: can you add a short description why those are needed

**Review by datapythonista — CHANGES_REQUESTED** [review]
I think we should test this in windows, I guess `Makefile` is rarely used if ever, but we want to make sure the environment can be resolved.

I'll see if I have time to add the build of the docs from windows, and then we'll have a windows build using `environment.yml`, but if someone can test this before manually that would be great.

File: environment.yml (hunk 2 above)
Comment (marcelotrevisani) [reply environment.yml]: Sure, I will rebase mu branch from master and that should be reverterd

File: environment.yml (hunk 3 above)
Comment (marcelotrevisani) [reply environment.yml]: Sure! I will add the explanation, actually the ``zlib`` we can remove since it is already a dependency of other packages

Comment (marcelotrevisani) [reply environment.yml]: I also think that the ``conda-forge`` channel should be prioritized by the defaults
since not all of the packages are present on the default channel and the project is using the missing packages from ``conda-forge``, you might see some problems in the future (unlikely) but it is better to avoid any kind of problem

But I can create another PR for that

**Comment by marcelotrevisani** [comment]
some CI are failing in the test phase
```
  File "<string>", line 1, in <module>
  File "/home/vsts/work/1/s/pandas/__init__.py", line 171, in <module>
    "{input_files_cla}").format(**locals())
KeyError: 'self'
```

**Comment by marcelotrevisani** [comment]
I will squash my commits when it is ready to be integrated.
I just found one problem with Linux, for some reason one of the builds is still using the compiler from the system. It needs more investigation

**Comment by TomAugspurger** [comment]
> I will squash my commits when it is ready to be integrated.

No need to squash, we do that on merge.

**Comment by marcelotrevisani** [comment]
> @marcelotrevisani is this still active? Can you merge master?

Sure, during this weekend I will have some time and I am going to do it

File: ci/setup_env.sh [hunk 4]
```diff
@@ -103,6 +81,46 @@ fi
...
+    ccache --version
+    
+    if [ -z "$GCC" ]; then
+        GCC=$(which gcc)
+    fi
+    echo "gcc: $GCC"
+    CCACHE=ccache
+    echo "ccache: $CCACHE"
+    export CC="ccache $CC"
+elif [ -z "$NOCACHE" ] && [ "${TRAVIS_OS_NAME}" == "osx" ]; then
```
Comment (WillAyd) [inline ci/setup_env.sh]: Can this conditional branch be combined with the one above? Seems like mostly copy / paste save one off things so would be easier to read if consolidated

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
@marcelotrevisani can you rebase

**Comment by jreback** [comment]
@marcelotrevisani we have had a fair amount of CI churn recently, so if you can merge master we can look again

**Comment by WillAyd** [comment]
Nice PR but I think this has gone stale. @marcelotrevisani ping if you'd like to pick this back up
